File: cpp-build/test/mileage_claim_collection_builder.cpp
#include "mileage_claim_collection_builder.hpp"

namespace shouty_report
{
const std::vector<mileage_claim>
mileage_claim_collection_builder::default_mileage_claim_collection = {
      mileage_claim("David Allen", 130000, 57),
      mileage_claim("Lisa Crispin", 27000, 19),
      mileage_claim("Ian Dees", 19238855, 22)
};

std::vector<mileage_claim> mileage_claim_collection_builder::build() const
{
    if (mileage_claim_collection.size() == 0)
    {
        return default_mileage_claim_collection;
    }
    else
    {
        return mileage_claim_collection;
    }
}

mileage_claim_collection_builder &
mileage_claim_collection_builder::with_sales_person_miles_for_customer(
    const std::string & name,
    const int miles,
    const int id
)
{
    mileage_claim_collection.push_back(mileage_claim(name, miles, id));
    return *this;
}

}

File: csharp-vs/ShoutyReportJobTest/Builders/MileageClaimCollectionBuilder.cs
using System.Collections.Generic;

namespace ShoutyReport.Builders
{
    public class MileageClaimCollectionBuilder
    {
        private List<MileageClaim> defaultMileageClaimCollection =
            new List<MileageClaim>(){
                new MileageClaim("David Allen", 130000, 57),
                new MileageClaim("Lisa Crispin", 27000, 19),
                new MileageClaim("Ian Dees", 19238855, 22)
            };

        private List<MileageClaim> mileageClaimCollection = null;

        public List<MileageClaim> build() {
            if (mileageClaimCollection == null) {
                return defaultMileageClaimCollection;
            } else {
                return mileageClaimCollection;
            }
        }

        public MileageClaimCollectionBuilder withSalesPersonMilesForCustomer(
          string name,
          int miles,
          int id) {

            if (mileageClaimCollection == null){
                mileageClaimCollection = new List<MileageClaim>();
            }

            mileageClaimCollection.Add(new MileageClaim(name, miles, id));
            return this;
        }
    }
}
